### PR TITLE
Fix wrong ignore path and create a lot if features

### PR DESCRIPTION
**Pull request by ASMfreaK**
This PR fixes the issue of wrong import path for the engine. The fix is simple : we put everything needed for the wasm script to run in a single library, hosted on crates.io. (This includes former `engine` and `shared` crates).

I've also added a few features for overall more flexible usage:
1. Added a flags to output different stages of the wasm module build, an option to rename the module, etc.
2. The only **breaking change** is in the output naming. The default name of the module is now "tinysearch-implementation".
3. I've added an ability to use tinysearch as a library in a bigger wasm project directly. There is an example on how to do it in the new `examples` folder.

### CODE REVIEW

File: tinysearch/assets/crate/Cargo.toml [hunk 1]
```diff
@@ -1,5 +1,5 @@
 [package]
-name = "tinysearch-engine"
+name = "THIS_VALUE_SHOULD_BE_FILLED"
```
Comment (mre) [inline tinysearch/assets/crate/Cargo.toml:2]: I'm missing the part where these values will be filled. Where will that happen?

**Comment by mre** [comment]
Looks great! I like the the concept of stages; nice idea.

Couple of comments:

* [ ] We need to fix the tests to make the build pass. Right now it fails on `include_str!("../assets/demo.html");`.
* [ ] Added a comment about `THIS_VALUE_SHOULD_BE_FILLED`, which I'm a bit confused by
* [ ] What was the reason behind changing the name of the output module? I don't mind too much, but maybe we can rethink that decision if it's just for aesthetical reasons to avoid a breaking change. If we want to go forward with the rename, then we could think about different names as well, such as `tinysearch-output`.

Thanks a lot for looking into this.

File: tinysearch/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,410 @@
...
+        cargo_toml_contents["package"]["name"] = value(self.crate_name.clone());
+        cargo_toml_contents["dependencies"]["tinysearch-engine"] =
+            toml_edit::Item::Table(self.engine_version.clone());
+        if self.non_top_level{
+            cargo_toml_contents.as_table_mut().remove("workspace");
+            cargo_toml_contents.as_table_mut().remove("profile");
+            cargo_toml_contents.as_table_mut().remove("lib");
+            cargo_toml_contents["lib"] = toml_edit::table();
+        }
+        fs::write(&cargo_toml, cargo_toml_contents.to_string())?;
```
Comment (ASMfreaK) [inline tinysearch/src/main.rs:284]: This is where the values in the `Cargo.toml` are set

**Comment by ASMfreaK** [comment]
>     * [ ]  We need to fix the tests to make the build pass. Right now it fails on `include_str!("../assets/demo.html");`.

I'll look into this, right now I can't reproduce the issue locally.

>     * [ ]  Added a comment about `THIS_VALUE_SHOULD_BE_FILLED`, which I'm a bit confused by

I've added a comment here oh GH, I'll add a comment in `Cargo.toml` template. You see, the file is no longer copied as is, but it is now a template, which is filled out in the generating code.

>     * [ ]  What was the reason behind changing the name of the output module? I don't mind too much, but maybe we can rethink that decision if it's just for aesthetical reasons to avoid a breaking change. If we want to go forward with the rename, then we could think about different names as well, such as `tinysearch-output`.

The reason is like this: since the merger of two libs (`shared` and `engine`) in one crate, I gave it `tinysearch-engine` name. Therefore the generated crate needed to be renamed to something other, than `tinysearch-engine`.
I now think, that the shared crate should retain its name - `tinysearch-shared`. I'll rename it back and bump all the versions to `0.8.0`. This way the output could remain named `tinysearch-engine`.

File: tinysearch/src/main.rs (hunk 2 above)
Comment (mre) [reply tinysearch/src/main.rs:284]: Makes perfect sense, thanks.

**Comment by mre** [comment]
> I'll look into this, right now I can't reproduce the issue locally.

Didn't you remove `bin/assets/demo.html`? I can't see that it was moved in your PR. That would explain the test failure, but not why it doesn't fail locally. 

> I now think, that the shared crate should retain its name - tinysearch-shared. I'll rename it back and bump all the versions to 0.8.0. This way the output could remain named tinysearch-engine.

Cool!

**Comment by ASMfreaK** [comment]
Here is a new version.
I've merged lib and bin crates into one - now it's called `tinysearch`. We should bump at least minor version.
The installation is now done:

```
cargo install --features=bin tinysearch
```

I hope, we'll see why it failed earlier with `demo.html`

**Comment by mre** [comment]
Guess we also need to update the CI pipelines; specifically the `publish-check`. Not sure what's up with the `build` pipeline, but maybe we can ignore that for now. As for `publish-check`, the paths to the crates need to be updated. Can you have a look?

**Comment by ASMfreaK** [comment]
It seems that we need to replace actions-rs clippy implemebration with [like here](https://github.com/groszewn/tensorboard/commit/73958a5774fdc8e605c4bbb66a7f3a1174482200) with another, better maintained one.

**Comment by ASMfreaK** [comment]
Can you please take a look?

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -38,33 +48,74 @@ jobs:
...
+        run: cargo install --force --path tinysearch --features=bin
 
       - name: Build WASM package from test index
-        run: RUST_LOG=debug tinysearch fixtures/index.json
+        run: RUST_LOG=debug tinysearch -e 'path= "'$PWD'/tinysearch"' fixtures/index.json
+      
+      - name: Build examples
+        run:
+          for example in `ls -1 examples`; do
+            trunk build examples/$example/index.html || exit 1
```
Comment (mre) [inline .github/workflows/ci.yml:117]:
```suggestion
            trunk build examples/$example/index.html || exit 1;
```

**Comment by Jieiku** [comment]
@ASMfreaK I just tried building this from git and then using it on a sample json file unsuccessfully: https://github.com/tinysearch/tinysearch/issues/151#issuecomment-1508982107

**Comment by mre** [comment]
@ASMfreaK thanks for your contribution. I made a few more changes and this finally looked good to go, so I merged it. I tested it locally before and didn't run into any issues. @Jieiku fyi.

**Pull request merged**